### DISCUSSION TITLE
Flows support for saving binary data as a file

### DISCUSSION THREAD

**Discussion by JSDA123** [directus/directus#23555] [Feature Requests]
### Summary

Add support in Flows to save binary data as a file in Directus that is obtained from a third-party APIs where the binary data is included in the response body. 

### Basic Example

In Flows, make a webhook POST to an external API and receive the response body which includes a file in binary format. The new feature would allow using a subsequent operation to save the file in Directus by referencing the binary data, for example as {{$last.data}}.  

### Motivation

The ability to work with binary data could apply to many use cases, especially PDF generation, for example creating a PDF invoice.  This use case wants to use flows to send a data array (or an HTML document) to an external PDF generation API (such as Stirling PDF).  The external API returns a generated PDF in the response body in binary format. The new feature would allow handling of the received binary data and save it as a file in directus.

Attempting to use a flows webhook to the /files endpoint to create a new file from the binary data results in the error: "Unexpected end of form".  Assuming I had not grasped the documentation, I opened a ticket for help and was informed by @rijkvanzanten this is not supported and would be a new feature:  https://github.com/directus/directus/issues/23361#event-14107483605

### Detailed Design

The ability to directly create a new file using binary data via a webhook to /files would be sufficient. 

However since this could be a common requirement, and using REST to create a file is a bit complicated, it might be ideal to crate a new flow operation type "Add File" could handle the details, perhaps with options to select the default folder and set some field data.

### Must Have:
- Flows ability to save binary data as a file, ie: {{$last.data}} --> directus file

### Could Have:
- a new "Add File" or "Add Binary File" operation that provides a simple GUI
- options to directly download the file in data studio instead of saving, for example a manually triggered flow 

### Drawbacks

I can't think of any drawbacks ...

### Alternatives

Many use cases revolve around PDF generation, so as an alternative the ability for Directus to handle PDF generation natively, or via an official extension, without the need for third-party APIs.  This would address my requirements, but might not be sufficient for others who need to work with binary data in flows.

### Adoption Strategy

The new feature would be included in a subsequent release or made available via an official extension.

**Comment by rafarc21** [comment]
This would be extremely useful to us. It's quite common for APIs to do this.

**Comment by dnl13** [comment]
I'm honestly shocked to find out this is a known limitation. How has this not been implemented yet? Handling file uploads via Flows is an absolutely critical requirement for many business cases.

What is the underlying issue? Is it the abstraction layer for the different storage drivers?

This request has been open for over a year and a half. It is hard to grasp that this is not prioritized as a necessary feature of a Headless CMS.

Since this is "not planned" currently: Is there at least ANY extension in the Marketplace that patches this gap so we can process files within Flows?

**Comment by domdus** [comment]
I did not test it but I think this extension operation can handle it. 
[https://directus.io/extensions/directus-extension-downloadstorefile-operation](https://directus.io/extensions/directus-extension-downloadstorefile-operation)

(Available in market place)
